### DISCUSSION TITLE
PSA: Type-privacy

### DISCUSSION THREAD

**Discussion by thomaseizinger** [libp2p/rust-libp2p#4632] [General]
For a while now, we've been making many modules in our plugins private to be able to "freely" modify them. This especially concerns implementations of `ConnectionHandler` and the events around it.

Well, I've got bad news: They are not actually private.

It is completely legal to write this today[^1]:

```rust
let event = <<<Behaviour<MemoryStore> as NetworkBehaviour>::ConnectionHandler as ConnectionHandler>::ToBehaviour>::ProtocolNotSupported { endpoint: ConnectedPoint::Dialer { address: Multiaddr::empty(), role_override: Endpoint::Dialer } };
```

And bam, you've constructed yourself an event that is only meant to be sent by the `ConnectionHandler` to the `NetworkBehaviour`. Not sure what to do about this yet but I am leaning towards documenting that these aren't actually public API and doing these kind of shenanigans is setting yourself up for bad times.

[^1]: On nightly. It needs the `#![feature(more_qualified_paths)]` but I am assuming this will stabilise eventually.

**Comment by mxinden** [comment]
> And bam, you've constructed yourself an event that is only meant to be sent by the `ConnectionHandler` to the `NetworkBehaviour`. Not sure what to do about this yet but I am leaning towards documenting that these aren't actually public API and doing these kind of shenanigans is setting yourself up for bad times.

:+1: sounds good to me.
